main.go

```diff
@@ -124,70 +183,91 @@ func exec(logger log.Logger, reg *prometheus.Registry, opts options) error {
+	flag.StringVar(&cfg.logFormat, "log.format", internal.LogFormatLogfmt,
 		"The log format to use. Options: 'logfmt', 'json'.")
-	flag.StringVar(&opts.listen, "web.listen", ":8080",
+	flag.StringVar(&cfg.listen, "web.listen", ":8080",
 		"The address on which internal server runs.")
-	flag.DurationVar(&opts.gracePeriod, "web.grace-period", server.DefaultGracePeriod,
+	flag.DurationVar(&cfg.gracePeriod, "web.grace-period", server.DefaultGracePeriod,
 		"The time to wait after an OS interrupt received.")
-	flag.DurationVar(&opts.timeout, "web.timeout", server.DefaultTimeout,
+	flag.DurationVar(&cfg.timeout, "web.timeout", server.DefaultTimeout,
 		"The maximum duration before timing out the request, and closing idle connections.")
 	flag.StringVar(&rawMetricsReadEndpoint, "metrics.read.endpoint", "",
 		"The endpoint against which to send read requests for metrics. It used as a fallback to 'query.endpoint' and 'query-range.endpoint'.")
 	flag.StringVar(&rawMetricsUIEndpoint, "metrics.ui.endpoint", "",
 		"The endpoint which forward ui requests.")
 	flag.StringVar(&rawMetricsWriteEndpoint, "metrics.write.endpoint", "",
 		"The endpoint against which to make write requests for metrics.")
-	flag.IntVar(&opts.proxyBufferCount, "proxy.buffer-count", proxy.DefaultBufferCount,
+	flag.IntVar(&cfg.proxy.bufferCount, "proxy.buffer-count", proxy.DefaultBufferCount,
 		"Maximum number of of reusable buffer used for copying HTTP reverse proxy responses.")
-	flag.IntVar(&opts.proxyBufferSizeBytes, "proxy.buffer-size-bytes", proxy.DefaultBufferSizeBytes,
+	flag.IntVar(&cfg.proxy.bufferSizeBytes, "proxy.buffer-size-bytes", proxy.DefaultBufferSizeBytes,
 		"Size (bytes) of reusable buffer used for copying HTTP reverse proxy responses.")
-	flag.DurationVar(&opts.proxyFlushInterval, "proxy.flush-interval", proxy.DefaultFlushInterval,
+	flag.DurationVar(&cfg.proxy.flushInterval, "proxy.flush-interval", proxy.DefaultFlushInterval,
 		"The flush interval to flush to the proxy while copying the response body. If zero, no periodic flushing is done. "+
 			"A negative value means to flush immediately after each write to the client.")
+	flag.StringVar(&cfg.tls.certFile, "tls-cert-file", "",
+		"File containing the default x509 Certificate for HTTPS. Leave blank to disable TLS.")
+	flag.StringVar(&cfg.tls.keyFile, "tls-private-key-file", "",
+		"File containing the default x509 private key matching --tls-cert-file. Leave blank to disable TLS.")
+	flag.StringVar(&cfg.tls.clientCAFile, "tls-client-ca-file", "",
+		"File containing the TLS CA to verify clients against."+
+			"If no client CA is specified, there won't be any client verification on server side.")
+	flag.StringVar(&cfg.tls.minVersion, "tls-min-version", "VersionTLS13",
+		"Minimum TLS version supported. Value must match version names from https://golang.org/pkg/crypto/tls/#pkg-constants.")
+	flag.StringVar(&rawTLSCipherSuites, "tls-cipher-suites", "",
```

Comment:
I think we've done some previous research on sane defaults for these, but I think it was based around TLS 1.2, maybe we can do this again for 1.3?

Reply:
👍 Let me look into it.

Reply:
Apparently, we cannot even set the `CipherSuite` if we use TLS 1.3. I'm attaching the related documentation. So sane default should be `nil`.

https://golang.org/pkg/crypto/tls/#CipherSuite

```diff
CipherSuites is a list of supported cipher suites for TLS versions up to TLS 1.2. 
If CipherSuites is nil, a default list of secure cipher suites is used, 
with a preference order based on hardware performance. 
The default cipher suites might change over Go versions. 
+Note that TLS 1.3 ciphersuites are not configurable.
```

Reply:
Amazing. I didn't even know.